### PR TITLE
DROOLS-2992 : Can't compile the kie module which has guided scorecard

### PR DESCRIPTION
**Pull request by Rikkola**
https://issues.jboss.org/browse/DROOLS-2992

https://github.com/kiegroup/drools/pull/2366
https://github.com/kiegroup/drools-wb/pull/1153

### CODE REVIEW

**Comment by Rikkola** [comment]
@jomarko Draft PRs for the build. I'll be adding tests in to this PR tonight or tomorrow morning.

File: drools-workbench-models/drools-workbench-models-guided-scorecard/src/main/java/org/drools/workbench/models/guided/scorecard/backend/ImportsToFQCN.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+import java.util.List;
+
+import org.kie.soup.project.datamodel.imports.Import;
+
+public class ImportsToFQCN {
+
+    private List<Import> imports;
+
+    public ImportsToFQCN(List<Import> imports) {
```
Comment (jomarko) [inline drools-workbench-models/drools-workbench-models-guided-scorecard/src/main/java/org/drools/workbench/models/guided/scorecard/backend/ImportsToFQCN.java]: Could we please be consistent in using `final` ? `final List<Import> imports`

File: drools-workbench-models/drools-workbench-models-guided-scorecard/src/main/java/org/drools/workbench/models/guided/scorecard/backend/GuidedScoreCardModelVisitor.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        return fullyQualifiedClassNames;
+    }
+
+    public Set<String> getConsumedModelClasses() {
+        //Convert Fact Types into Fully Qualified Class Names
+        final Set<String> fullyQualifiedClassNames = new HashSet<String>();
+        for (String factType : getFactTypes()) {
+            fullyQualifiedClassNames.add(convertToFullyQualifiedClassName(factType));
```
Comment (jomarko) [inline drools-workbench-models/drools-workbench-models-guided-scorecard/src/main/java/org/drools/workbench/models/guided/scorecard/backend/GuidedScoreCardModelVisitor.java:68]: Don't we need ` if (factType.contains(".")) {` ?

File: drools-workbench-models/drools-workbench-models-guided-scorecard/src/main/java/org/drools/workbench/models/guided/scorecard/backend/GuidedScoreCardModelVisitor.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        factTypes.add(model.getFactName());
+        for (Characteristic c : model.getCharacteristics()) {
+            factTypes.add(c.getFact());
+        }
+        return factTypes;
+    }
+
+    //Get the fully qualified class name of the fact type
+    private String convertToFullyQualifiedClassName(final String factType) {
+        if (factType.contains(".")) {
```
Comment (jomarko) [inline drools-workbench-models/drools-workbench-models-guided-scorecard/src/main/java/org/drools/workbench/models/guided/scorecard/backend/GuidedScoreCardModelVisitor.java:86]: Oh, so we can remove the if from `getConsumedPackageImports` I think

File: drools-workbench-models/drools-workbench-models-guided-scorecard/src/main/java/org/drools/workbench/models/guided/scorecard/backend/GuidedScoreCardModelVisitor.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        String fullyQualifiedClassName = null;
+        for (Import imp : imports.getImports()) {
+            if (imp.getType().endsWith(factType)) {
+                fullyQualifiedClassName = imp.getType();
+                break;
+            }
+        }
+        if (fullyQualifiedClassName == null) {
+            fullyQualifiedClassName = packageName + "." + factType;
+        }
```
Comment (jomarko) [inline drools-workbench-models/drools-workbench-models-guided-scorecard/src/main/java/org/drools/workbench/models/guided/scorecard/backend/GuidedScoreCardModelVisitor.java]: I think int this case `Streams`, `filter`, `orElse` would make it more readable.

**Review by jomarko — COMMENTED** [review]
Some comments to code

File: drools-workbench-models/drools-workbench-models-guided-scorecard/src/main/java/org/drools/workbench/models/guided/scorecard/backend/GuidedScoreCardModelVisitor.java (hunk 2 above)
Comment (Rikkola) [reply drools-workbench-models/drools-workbench-models-guided-scorecard/src/main/java/org/drools/workbench/models/guided/scorecard/backend/GuidedScoreCardModelVisitor.java:68]: `getConsumedPackageImports` should ignore all the imports that have a dot and `getConsumedModelClasses` should contain every class. I'll rename the methods or write JavaDocks to explaing this.

File: drools-workbench-models/drools-workbench-models-guided-scorecard/src/main/java/org/drools/workbench/models/guided/scorecard/backend/GuidedScoreCardModelVisitor.java (hunk 4 above)
Comment (Rikkola) [reply drools-workbench-models/drools-workbench-models-guided-scorecard/src/main/java/org/drools/workbench/models/guided/scorecard/backend/GuidedScoreCardModelVisitor.java]: I cleaned this area up a bit, hopefully making it more readable.

Comment (jomarko) [reply drools-workbench-models/drools-workbench-models-guided-scorecard/src/main/java/org/drools/workbench/models/guided/scorecard/backend/GuidedScoreCardModelVisitor.java]: +1

**Comment by Rikkola** [comment]
Asked @lanceleverich for code review since he was the one who first investigated this issue.

**Review by jomarko — APPROVED** [review]

File: drools-workbench-models/drools-workbench-models-guided-scorecard/src/test/java/org/drools/workbench/models/guided/scorecard/backend/base/GuidedScoreCardDRLPersistenceTest.java [hunk 5]
```diff
@@ -117,9 +119,11 @@ public void testModelWithImportsAndFactName() {
                      StringUtil.countMatches(drl,
                                              "rule \""));
 
-        assertEquals(14,
```
Comment (lanceleverich) [inline drools-workbench-models/drools-workbench-models-guided-scorecard/src/test/java/org/drools/workbench/models/guided/scorecard/backend/base/GuidedScoreCardDRLPersistenceTest.java:120]: I think we should stay away from these "magic number" based tests. It seems that a simple change to how the rules are generated, or to the underlying templates, could cause these to break.

Comment (Rikkola) [reply drools-workbench-models/drools-workbench-models-guided-scorecard/src/test/java/org/drools/workbench/models/guided/scorecard/backend/base/GuidedScoreCardDRLPersistenceTest.java:120]: True, better to test for content like after this line that I had to change.

Comment (Rikkola) [reply drools-workbench-models/drools-workbench-models-guided-scorecard/src/test/java/org/drools/workbench/models/guided/scorecard/backend/base/GuidedScoreCardDRLPersistenceTest.java:120]: Removed.

**Review by lanceleverich — APPROVED** [review]
Looks good now that we have removed the "magic number" based test

**Pull request merged**